physicsnemo/distributed/shard_utils/attention_patches.py

```diff
@@ -0,0 +1,695 @@
+# SPDX-FileCopyrightText: Copyright (c) 2023 - 2024 NVIDIA CORPORATION & AFFILIATES.
+# SPDX-FileCopyrightText: All rights reserved.
+# SPDX-License-Identifier: Apache-2.0
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from typing import Any, Callable, Optional, Tuple, Union
+
+import torch
+import torch.distributed as dist
+from torch.autograd.profiler import record_function
+
+from physicsnemo.utils.version_check import check_module_requirements
+
+check_module_requirements("physicsnemo.distributed.shard_tensor")
+
+
+from torch.distributed import DeviceMesh  # noqa: E402
+
+from physicsnemo.distributed import ShardTensor  # noqa: E402
+from physicsnemo.distributed.shard_utils.patch_core import (  # noqa: E402
+    MissingShardPatch,
+)
+from physicsnemo.distributed.shard_utils.ring import (  # noqa: E402
+    RingPassingConfig,
+    perform_ring_iteration,
+)
+
+aten = torch.ops.aten
+
+
+def add_log_sumexp(
+    log_a: Optional[torch.Tensor], log_b: Optional[torch.Tensor]
+) -> torch.Tensor:
```

Comment:
Notably, this function currently returns None if both log_a and log_b are None, which the type hints indicate is permissible. I think we should error if both are None, rather than allow a later error when a variable expecting a tensor becomes None.

Reply:
Yes, but deferring.